src/arch/src/aarch64/gic/redist_regs.rs

```diff
@@ -0,0 +1,207 @@
+const GICR_ICACTIVER0: u32 = GICR_SGI_OFFSET + 0x0380;
+const GICR_IPRIORITYR0: u32 = GICR_SGI_OFFSET + 0x0400;
+const GICR_ICFGR0: u32 = GICR_SGI_OFFSET + 0x0C00;
+
+const KVM_DEV_ARM_VGIC_V3_MPIDR_SHIFT: u32 = 32;
+const KVM_DEV_ARM_VGIC_V3_MPIDR_MASK: u64 = 0xffff_ffff << KVM_DEV_ARM_VGIC_V3_MPIDR_SHIFT as u64;
+
+// All or at least the registers we are interested in are 32 bit, so
+// we use a constant for size(u32).
+const REG_SIZE: u8 = 4;
+
+/// This is how we represent the registers of a distributor.
+/// It is relrvant their offset from the base address of the
+/// distributor.
+/// Each register has a different number
+/// of bits_per_irq and is therefore variable length.
+/// First 32 interrupts (0-32) are private to each CPU (SGIs and PPIs).
+/// and so we save the first irq to identify between the type of the interrupt
+/// that the respective register deals with.
+struct RdistReg {
+    /// Offset from distributor address.
+    base: u32,
+    /// Length of the register.
+    length: u8,
+}
+
+// Creates a vgic redistributor register.
+macro_rules! VGIC_RDIST_REG {
+    ($base:expr, $len:expr) => {
+        RdistReg {
+            base: $base,
+            length: $len,
+        }
+    };
+}
+
+// List with relevant distributor registers that we will be restoring.
+static VGIC_RDIST_REGS: &[RdistReg] = &[
+    VGIC_RDIST_REG!(GICR_CTLR, 4),
+    VGIC_RDIST_REG!(GICR_STATUSR, 4),
+    VGIC_RDIST_REG!(GICR_WAKER, 4),
+    VGIC_RDIST_REG!(GICR_PROPBASER, 8),
+    VGIC_RDIST_REG!(GICR_PENDBASER, 8),
+];
+
+// List with relevant distributor registers that we will be restoring.
+static VGIC_SGI_REGS: &[RdistReg] = &[
+    VGIC_RDIST_REG!(GICR_IGROUPR0, 4),
+    VGIC_RDIST_REG!(GICR_ICENABLER0, 4),
+    VGIC_RDIST_REG!(GICR_ISENABLER0, 4),
+    VGIC_RDIST_REG!(GICR_ICFGR0, 8),
+    VGIC_RDIST_REG!(GICR_ICPENDR0, 4),
+    VGIC_RDIST_REG!(GICR_ISPENDR0, 4),
+    VGIC_RDIST_REG!(GICR_ICACTIVER0, 4),
+    VGIC_RDIST_REG!(GICR_ISACTIVER0, 4),
+    VGIC_RDIST_REG!(GICR_IPRIORITYR0, 32),
+];
+
+#[allow(clippy::trivially_copy_pass_by_ref)]
+fn redist_attr_access(fd: &DeviceFd, offset: u32, typer: u64, val: &u32, set: bool) -> Result<()> {
+    let mut gic_dist_attr = kvm_device_attr {
+        group: KVM_DEV_ARM_VGIC_GRP_REDIST_REGS,
+        attr: (typer & KVM_DEV_ARM_VGIC_V3_MPIDR_MASK) | (offset as u64), // this needs the mpidr
+        addr: val as *const u32 as u64,
+        flags: 0,
+    };
+    if set {
+        fd.set_device_attr(&gic_dist_attr)
+            .map_err(|e| Error::DeviceAttribute(e, true, KVM_DEV_ARM_VGIC_GRP_REDIST_REGS))?;
+    } else {
+        fd.get_device_attr(&mut gic_dist_attr)
+            .map_err(|e| Error::DeviceAttribute(e, false, KVM_DEV_ARM_VGIC_GRP_REDIST_REGS))?;
+    }
+    Ok(())
+}
+
+fn access_redists_aux(
+    fd: &DeviceFd,
+    gicr_typer: &[u64],
+    state: &mut Vec<u32>,
+    reg_list: &'static [RdistReg],
+    idx: &mut usize,
+    set: bool,
```

Comment:
I noticed a couple of issues here:
1. `idx` is only needed for set
2. state only needs to be mutable for get, but we always expect it to be `&mut`. Because of this, inside `set_redist_regs` we have to do `state.to_owned()` which AFAIU clones the data leading to an overhead.
3. I find `set: bool` a bit hard to read and easy to get wrong.

I would like to propose instead of `set: bool` something like 

```
enum Action<'a> {
    Set(&'a [u32], usize),
    Get(&'a mut Vec<u32>),
}
```
I'm not sure about the name. `Action` is not very suggestive. Anyway, then we can do:

```
fn access_redists_aux(
    fd: &DeviceFd,
    gicr_typer: &[u64],
    reg_list: &'static [RdistReg],
    action: &mut Action,
) -> Result<()> {
    for i in gicr_typer {
        for rdreg in reg_list {
            let mut base = rdreg.base;
            let end = base + rdreg.length as u32;

            while base < end {
                let mut val = 0;
                match action {
                    Action::Set(state, idx) => {
                        val = state[*idx];
                        redist_attr_access(fd, base, *i, &val, true)?;
                        *idx += 1;
                    }
                    Action::Get(state) => {
                        redist_attr_access(fd, base, *i, &val, false)?;
                        state.push(val);
                    }
                }
                base += REG_SIZE as u32;
            }
        }
    }
    Ok(())
}

/// Get redistributor registers.
pub fn get_redist_regs(fd: &DeviceFd, gicr_typer: &[u64]) -> Result<Vec<u32>> {
    let mut state = Vec::new();
    let mut action = Action::Get(&mut state);
    access_redists_aux(
        fd,
        &gicr_typer,
        VGIC_RDIST_REGS,
        &mut action,
    )?;

    access_redists_aux(
        fd,
        &gicr_typer,
        VGIC_SGI_REGS,
        &mut action,
    )?;
    Ok(state)
}

/// Set redistributor registers.
pub fn set_redist_regs(fd: &DeviceFd, gicr_typer: &[u64], state: &[u32]) -> Result<()> {
    let mut action = Action::Set(state, 0);

    access_redists_aux(
        fd,
        gicr_typer,
        VGIC_RDIST_REGS,
        &mut action,
    )?;
    access_redists_aux(
        fd,
        gicr_typer,
        VGIC_SGI_REGS,
        &mut action,
    )
}
```